Allow picking from all workflow versions

johnbradley opened the pull request:
Lets users pick from all versions of a workflow (instead of just the latest version).
Fixes #54

## Select Workflow Screenshot
Shows the latest version of each workflow.
![screen shot 2017-08-25 at 9 32 26 am](https://user-images.githubusercontent.com/1024463/29716225-91d4352e-8978-11e7-9b8a-c43c6f34a6fa.png)

## Select Workflow Version Screenshot
User is sent here if the click 'Browse All Versions' from the Select Workflow page
![screen shot 2017-08-25 at 9 32 37 am](https://user-images.githubusercontent.com/1024463/29716231-935b6804-8978-11e7-84d0-447c0fa617d4.png)

johnbradley commented:
I thought about moving the tiny `version #` to the grey title bar above atop each selectable item.
This would require more changes to item-chooser but would make the two screens more consistent.
Something like this for the `Select Workflow` Page:
![blah](https://user-images.githubusercontent.com/1024463/29716737-82125b32-897a-11e7-92f1-92b3675d7d10.png)
And then something this for the `Select Workflow Version` Page
![blah2](https://user-images.githubusercontent.com/1024463/29716795-b701438a-897a-11e7-8864-b3d7a4440431.png)

dleehr commented:
I agree it would be better to have the version number in that title bar. I don't think we need to repeat the title there, just the version.

johnbradley commented:
If I put the version number as the title the user cannot click it and get to the packed cwl file.

dleehr commented:
I also thought about updating the component in-place to make the previous versions selectable right inside the panel as a list.

1. Initially, component shows the latest version and a link to show previous versions
2. If user clicks the link to show previous versions, the list of previous versions is shown below
3. If user clicks a version number from the list of previous versions, the description and currently selected version are changed to indicate the selected previous version
4. When an older version is selected, this should be clear

![previous-versions-picker](https://user-images.githubusercontent.com/2752599/29717828-9b5fed94-897e-11e7-8932-dbd886b3616b.png)

johnbradley commented:
Wouldn't they need the description to tell which one they want?

dleehr commented:
> If I put the version number as the title the user cannot click it and get to the packed cwl file.

Assuming that's still important to have, we should be able to link it from somewhere in the details. My point was that the title bar is the most emphasized thing in the component, and if it's nearly identical for every component on the page (because the workflow name is the same for all versions), then it's less effective.

It's certainly possible this "item-chooser" generic component is too generic :)

dleehr commented:
> Wouldn't they need the description to tell which one they want?

Possibly - We could at least add the date next to the version number, and switching between version numbers would cause the selected description to show on the top.

But this was just a design idea, and I think that, absent any actual user feedback, I probably shouldn't spend too much time on this.

So let's just find a way to put the version number in the title.

johnbradley commented:
I added the version number to the title for the workflow version items. I removed the url for now since the raw packed workflow is probably a little cryptic for the end users.

dleehr commented:
> I removed the link to the cwl workflow for now since the raw packed workflow is probably a little cryptic for the end users.

Yeah I think this is the better move. Maybe we could update the list of workflows/workflow versions (see #58) to provide some more details about the workflow versions, like a link to the CWL.

johnbradley commented:
Something unexpected is the screen from #58 changed here and now is showing descriptions.
This might be due to the itemTitleKey change.

johnbradley commented:
Additional changes to allow users to see more details about the workflow versions:
- Add nested route to workflows: `/workflows/<id>/version/<id>`
- Add link to this new route from each workflow version as right aligned `v<number>` subtitle

dleehr commented:
Per our call this morning, this is still in progress right?

johnbradley commented:
Correct

dleehr commented on tests/unit/routes/workflows/versions/index-test.js:
```diff
@@ -0,0 +1,11 @@
+  // needs: ['controller:foo']
+});
+
+test('it exists', function(assert) {
```
I know many of the route tests are empty/trivial, but since this route has some dependencies on its nesting, it's probably good to test it in case we ever move it around.

johnbradley replied:
Not quite sure how to test this. The emberjs guide is focused on testing actions: https://guides.emberjs.com/v2.14.0/testing/testing-routes/

dleehr replied:
> Acceptance tests will likely provide better coverage for routes because routes are typically used to perform transitions and load data, both of which are tested more easily in full context rather than isolation.

I think they don't automatically generate acceptance tests for you, since they're larger in scope and have no obvious defaults per route.

But here, we do have a unit test that could test the functionality we have - looking up the workflow id and querying for versions based on that. The unit test could simply instantiate the route with a mocked store (and params), and assert that it the store is called with the workflow id in the params
```

dleehr commented on app/templates/components/card-panel.hbs:
```diff
@@ -0,0 +1,5 @@
+<div class="col-md-4">
```
Ember will already wrap this in a div, so if you don't want an extra div, you can put the outermost HTML element into the js:

```
...
tagName: "div",
classNames: ["col-md-4"],
...
```

johnbradley commented on app/components/jobs/workflow-card.js:
```diff
@@ -0,0 +1,13 @@
+
+export default Ember.Component.extend({
+  workflow: null,
+  onPicked: null,
```
I'm not following you here. workflow should be passed a workflow model not a function.
Also, why would you need a type hint for a named attribute vs a positional parameter?

dleehr replied:
The comment was for `onPicked`, not workflow, . When someone uses this component, they can't do it without providing a value for `onPicked`. If they don't the component tries to call `null` as a function.

So, I suggest putting in a default function implementation for `onPicked`, mostly as a hint to the user on what the type of this property should be, since it's otherwise hard to determine.

The named attribute vs positional parameter doesn't make any technical difference here, just that it's not obvious to someone using this code that they have to provide values for named attributes when using the component. Declaring positional parameters encourages providing values when using the component.

johnbradley commented on app/components/next-back-buttons.js:
```diff
@@ -0,0 +1,14 @@
+export default Ember.Component.extend({
+  onClickNext() {},
+  onClickBack() {},
+  actions: {
```
Could you provide an example?

dleehr replied:
This comment was about the action pattern we were discussing above.

johnbradley commented on app/templates/jobs/new/select-workflow-version.hbs:
```diff
@@ -0,0 +1,13 @@
+        <label class="control-label">Workflow version:</label>
+        <div class="row">
+          {{#each model.versions as |version| }}
+            {{jobs/workflow-version-card workflowVersion=version onPicked=(action 'setWorkflowVersion')}}
```
I tried this:
```
{{jobs/workflow-card workflow=workflow onPicked=setWorkflow}}
```
and got this:
```
ember.debug.js:19840 TypeError: onPicked is not a function
```

This variation also resulted in the same error:
```
{{jobs/workflow-card workflow=workflow onPicked=setWorkflow}}
```

Both reactjs and polymer don't require this syntax. But the emberjs docs use it: https://guides.emberjs.com/v2.14.0/components/handling-events/

johnbradley replied:
Details on the action 'helper': https://emberjs.com/api/ember/2.15.0/classes/Ember.Templates.helpers/methods/action?anchor=action

dleehr replied:
I tried some variations too - moving `setWorkflow` from `actions: {}` to be function or closure in the root of the `select-workflow-versions.js` component. Doing that, the hbs template can pass that closure by name (`onPicked=setWorkflow`) down to children it renders. 

But then, in the child, it's anonymous closure and `this` is not bound to the `select-workflow-version` component. Obviously there are ways to handle that but the pattern seems clunky.

I know promoting these functions to actions and sending them around does work, but the documentation on actions doesn't say much about this use case and it feels like too much code for something that should be easy.

This briefly shows our use case, but it's more meant to decode a drop event and call a parameter with it.
https://guides.emberjs.com/v2.14.0/components/handling-events/#toc_sending-actions

Some of the text there makes it sound like we should be able to name actions in parents and send them from children without additional linking. Or bind targets to actions. I'll test some of these ideas in a sample project

johnbradley commented on tests/integration/components/card-panel-test.js:
```diff
@@ -0,0 +1,25 @@
+
+  this.render(hbs`{{card-panel}}`);
+
+  assert.equal(this.$().text().trim(), '');
```
This is the entirety of card-panel:
```
<article class="card-panel panel panel-default">
  {{yield}}
</article>
```
What kind of tests did you have in mind?

dleehr replied:
Sorry, I was thinking of the things with actions :)
